source/v5/mqtt5_utils.c

```diff
@@ -537,3 +537,18 @@ bool aws_mqtt_is_topic_filter_shared_subscription(struct aws_byte_cursor topic_c
 
     return true;
 }
+
+// UTF-8 encoded string validation respect to [MQTT-1.5.3-2].
+bool aws_mqtt_utf8_validator(const uint32_t codepoint) {
+    // Control characters are not valid
+    if ((codepoint >= 0x0001 && codepoint <= 0x001F) || (codepoint >= 0x007F && codepoint <= 0x009F)) {
```

Comment:
A good optimizer would probably make these changes for you, but let's be careful and explicit on this since it's a per-character inner loop:

1. Remove the ( == 0) check on 549 and change the first if to just <= 0x1F
2. Change the  && 0xFFFF compound check to ((codepoint & 0xFFFF) >= 0xFFFE)

So

```
if (codepoint <= 0x1F) {
   return false;
}

if (codepoint >= 0x7F && codepoint <= 0x9F) {
   return false;
}

if ((codepoint & 0xFFFF) >= 0xFFFE) {
   return false;
}

if (codepoint >= 0xFDD0 && codepoint <= 0xFDEF) {
   return false;
}

return true;
```